pkg/volume/csi/csi_attacher.go

```diff
@@ -440,13 +464,22 @@ func (c *csiAttacher) waitForVolumeAttachDetachStatus(attach *storage.VolumeAtta
 		return nil
 	}
 
-	watcher, err := c.k8s.StorageV1().VolumeAttachments().Watch(context.TODO(), meta.SingleObject(meta.ObjectMeta{Name: attachID, ResourceVersion: attach.ResourceVersion}))
-	if err != nil {
-		return fmt.Errorf("watch error:%v for volume %v", err, volumeHandle)
+	// For AttachDetachVolumeHosts, VolumeAttachment watches are managed via the volumeAttachmentWatcher,
+	// which uses the CSI plugin's SharedInformer. For KubeletVolumeHosts, VolumeAttachment watches
+	// are managed via the API.
+	var ch <-chan watch.Event
+	if c.plugin.volumeAttachmentInformerHook != nil {
+		ch = c.plugin.volumeAttachmentInformerHook.GetWatch(attachID)
+	} else {
```

Comment:
I think the design and logic could be simpler if we didn't have to support kubelet attach/detach mode.

Unfortunately we don't test it so it's not clear if csi attach is even working in kubelet mode attach or not. I think it would fail mainly when trying to Create/Delete VolumeAttachment objects. I expect the node authorizer would reject it if it was coming from kubelet on a node. Can you try an experiment by enabling kubelet Attach/Detach to see if we do hit these errors? https://github.com/kubernetes/kubernetes/blob/87984d84d18c7816835e5efedebb17ed13aa2509/cmd/kubelet/app/options/options.go#L512

cc @jsafrane @saad-ali @gnufied @jingxu97  for thoughts.